### PR TITLE
lsp: support client side code action commands (vim.lsp.commands)

### PR DESCRIPTION
**Pull request by mfussenegger**
This builds on https://github.com/neovim/neovim/pull/14112 and closes
https://github.com/neovim/neovim/issues/12326

### CODE REVIEW

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp.lua:1317]: What would be the best way to get this into the lsp.txt docs?

**Comment by mfussenegger** [comment]
@mjlbach @tjdevries I think this might be something to get in for 0.5.0. It extends the `lsp-handler` interface, and although it only adds an argument, it still is a semi-breaking change. If a user or plugin currently extends a handler without passing all arguments via `...` to the original handler, then the new argument will be missing, which could cause issues.

In https://github.com/neovim/neovim/pull/14112 I also semi-seriously suggested that we could change the `lsp-handler` interface even more, but I suspect at this point there is already too much out there that would break and it would probably delay 0.5 further. On the other hand, once 0.5 is out it would be even harder to make any changes to it.

**Comment by mjlbach** [comment]
This seems fine/like a good idea to me, but I think TJ should be the one to sign-off on this.

File: runtime/lua/vim/lsp.lua
Comment (tjdevries) [inline runtime/lua/vim/lsp.lua:1337]: I'm thinking that I'd like to add things to `commands` via a function that can validate the function/object/whatever at registration time (not 100% sure though).

Another thing  to consider would be that some commands may want to be registered directly on the client, so we might want to provide a way to do that -- similar to how we do that with handlers.

**Comment by tjdevries** [comment]
One weird thing is this kind of breaks the symmetry of the `handlers` with them having different signatures now kind of. 

I think the docs need a bit more updating as well.

In regards to a larger breaking change to handlers, I would have probably made the same change or similar before, but it already felt quite well-established before I had a chance to do so in that PR. I can't think of any great way to make the change at this point that seems super worthwhile.

File: runtime/lua/vim/lsp.lua
Comment (tjdevries) [reply runtime/lua/vim/lsp.lua:1317]: it is hard currently :'( we basically only do doc generation for functions. Although my suggestion below would probably make it possible to do so.

Getting closer to getting the tree sitter parser to the point where we could actually do this.

File: runtime/lua/vim/lsp.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:1337]: > I'm thinking that I'd like to add things to commands via a function that can validate the function/object/whatever at registration time (not 100% sure though).


Added some basic validation logic
People could still override the commands completely, not sure if we want to prevent that as well?
`handlers` are currently entirely open too.

> Another thing to consider would be that some commands may want to be registered directly on the client, so we might want to provide a way to do that -- similar to how we do that with handlers.

Would you be okay to follow up on that once somebody asks for it? I think it would be a non-breaking extension

**Comment by mfussenegger** [comment]
> One weird thing is this kind of breaks the symmetry of the handlers with them having different signatures now kind of.

Yes I'm also not really happy with that. That's why I brought up the API change.

If you have any other suggestion how to bring the symmetry back I'm all ears.
I was even contemplating if it would be possible to utilize the inspection functions of `debug` to build some kind of compatibility layer between current signature and a changed signature. 
I think it might actually be possible, but not sure if we wanna go down that route?


> I think the docs need a bit more updating as well.

The docs for the `lsp-handler`, or somewhere else?

**Comment by kamalmarhubi** [comment]
@mfussenegger 
>  I was even contemplating if it would be possible to utilize the inspection functions of `debug` to build some kind of compatibility layer between current signature and a changed signature.
> I think it might actually be possible, but not sure if we wanna go down that route?

FWIW, I was able to use [this Stackoverflow snippet](https://stackoverflow.com/a/51102204) to introspect number of arguments expected by a function. This would let the LSP implementation change how it calls the handler / action functions to provide some backwards compatibility.

The extensible and symmetrical API seems pretty worthwhile. If you, @mjlbach, or @tjdevries think it's worth prototyping, I might give it a shot.

**Comment by mfussenegger** [comment]
@tjdevries @mjlbach any further inputs how to proceed with this?

I'd also like to add `codeAction/resolve` support after this - (see https://github.com/mfussenegger/nvim-jdtls/pull/74/files )

**Comment by tjdevries** [comment]
Comment from element:

```lua
M.wrap_05_handler = function(handler)
  return function(err, result, context)
    return handler(err, context.method, result, context.client_id, context.bufnr, context.config)
  end
end

-- Users can do:
local make_stuff = function(...)
  local handler = function(err, method, result, client_id, bufnr, config)
    do_things(...)
  end

  if vim.lsp.compat then
    handler = vim.lsp.compat.wrap_05_handler(handler)
  end

  return handler
end
```

I am ok with `function(err, method, context, config)` as well.

We can actually release vim.lsp.compat module _first_ with literally nothing (like it just returns the handler itself).

This would allow people to add that into their plugins for 0.6 compatible stuff and then we make the actual changes.

I'm hesitant to actually merge this after thinking more about it because I don't want 3 states of signatures (1. 0.5 release, 2. this PR, 3. context). I'd rather have only the 2 states

File: runtime/lua/vim/lsp.lua
Comment (justinmk) [inline runtime/lua/vim/lsp.lua:1549]: hmm, how does this address the use-case from #12326 :

> Currently a plugin could choose to override the `textDocument/codeAction` callback or provide a custom `code_action` function. But then each plugin that wants to handle some command in a custom way would end-up re-implementing the code-action boilerplate and would be incompatible with each other.

how will multiple plugins add hooks to the same code action?

Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:1549]: Both plugins would register their command handlers within `vim.lsp.commands`.
The command keys would still be unique. So you cannot have two plugins provide the same command. But plugins for different language server wouldn't conflict.

E.g nvim-jdtls would add entries like this:

```lua
vim.lsp.commands['java.apply.workspaceEdit'] = ...
vim.lsp.commands['java.action.organizeImports'] = ...
```

and rust-tools.nvim could add rust-analyzer specific entries:


```lua
vim.lsp.commands['rust-analyzer.runSingle'] = ...
```

**Comment by justinmk** [comment]
What's the rationale for naming it `vim.lsp.commands` instead of `vim.lsp.codeactions` ?

**Comment by mfussenegger** [comment]
> What's the rationale for naming it `vim.lsp.commands` instead of `vim.lsp.codeactions` ? Seems like `vim.lsp.commands` might be confused with other uses of the word "command" in various parts of Nvim.

`command` is used in the LSP specification (https://microsoft.github.io/language-server-protocol/specifications/specification-current/#command) and they're not limited to codeactions, but can also be invoked via code lenses.

Maybe `client_commands` , or `lsp_client_commands`? - although that feels redundant, given that it is already in the lsp module.

**Comment by justinmk** [comment]
makes sense. +1 for `vim.lsp.commands` then.

**Comment by mjlbach** [comment]
This PR can be updated/simplified now that https://github.com/neovim/neovim/commit/df17d7844ed7dedcb80f9405f7078a046a12524a is merged

File: runtime/lua/vim/lsp/handlers.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/handlers.lua:151]: I wonder if we should change `buf.execute_command` to do the `vim.lsp.commands` lookup, to ensure all lsp actions which deal with commands check the registry?

(We should also adapt it to use the right client, the `execute_command` here should not go through *all* clients if multiple are attached to the buffer).

File: runtime/lua/vim/lsp/handlers.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/handlers.lua:113]: I tried to keep the naming consistent as `result` for each handler and fully populate each handler with nil (`_`) args to make it obvious that they all have the same signature, up for discussion though.

File: runtime/lua/vim/lsp/handlers.lua
Comment (mjlbach) [reply runtime/lua/vim/lsp/handlers.lua:151]: I think I'm good with that, it makes sense. But is it the case a user might want to register special commands per client?

Comment (mfussenegger) [reply runtime/lua/vim/lsp/handlers.lua:151]: I think I've changed my mind a bit and I'd actually like to defer changing this to a later PR (where I either introduce `codeAction/resolve` support or where I integrate the commands with code-lenses.

In both cases it is necessary to use the right client to use the server side commands, so that will have an influence on the API

**Review by mfussenegger — COMMENTED** [review]
Rebased and added a fixup to rename the handler parameters.

File: runtime/doc/lsp.txt
Comment (mjlbach) [inline runtime/doc/lsp.txt:227]: This seems like a bad rebase

**Comment by mjlbach** [comment]
Same comment as before, but does it make sense to allow global command registration rather than doing something similar as with capabilties, and passing it to start_client (or both have global and per-client overrides)?

**Comment by mfussenegger** [comment]
> Same comment as before, but does it make sense to allow global command registration rather than doing something similar as with capabilties, and passing it to start_client (or both have global and per-client overrides)?

I kinda followed the global handler model - if we want to have client local overrides, we could still extend that later on without breaking changes if there are use cases.

**Comment by mjlbach** [comment]
Do you not anticipate name collisions across commands from different language servers? 

After the doc issue is fixed, this PR is good with me (it was the motivation for the handler change haha), but @justinmk reiterates his dislike of registration vs declaration last night on matrix, so I'll let him chime in first. I personally think this is the cleanest way and matches most of the style of overrides we do for handlers.

**Comment by mfussenegger** [comment]
> Do you not anticipate name collisions across commands from different language servers?

From what I've seen they all prefix the command names with either the language name or the language server name.

**Comment by mjlbach** [comment]
Nevermind, I misread the indentation on the documentation, so it's just pending commentary from Justin.

**Review by mjlbach — APPROVED** [review]

**Review by justinmk — APPROVED** [review]
LGTM.

>  registration vs declaration ... (I'm not sure, tbh, what the distinction would be in this case other than what I had mentioned about passing commands into start_client as an additional table)

For the Nvim LSP framework the architecture is already established; probably shouldn't introduce a new concept/pattern. "Active" registration of handlers is the common pattern in the Nvim LSP framework. 

"Passive" (declarative) plugin-like model minimizes "lifecycle" questions and tends to be more decoupled, so it's good to consider it first.

Thanks for considering :)

**Comment by weilbith** [comment]
Sorry if I missed something, but I just stumbled over it while working on weilbith/nvim-code-action-menu#10 to handle the just added code action resolve capability. Thereby I stumbled in the NeoVim Lua code about the part if the selected user code action/code lens is a user/plugin defined command in `vim.lsp.commands`. But I never see where this list is actually used to inject these entries into the lists the user can select from (in case of the code actions). Is this intended and comes later, did I missed it or...? 
Thanks for the help in advanced! 🙏🏾

**Comment by mfussenegger** [comment]
> But I never see where this list is actually used to inject these entries into the lists the user can select from (in case of the code actions). Is this intended and comes later, did I missed it or...?

The language server has to include them in the response (and whether they do is usually gated behind some capability flag that is set within the `settings` or `init_options` when the client is initialized).

For example, eclipse.jdt.ls supports a bunch of options like:

```
extendedClientCapabilities = {
  progressReportProvider = true;
  classFileContentsSupport = true;
  generateToStringPromptSupport = true;
  hashCodeEqualsPromptSupport = true;
  advancedExtractRefactoringSupport = true;
  advancedOrganizeImportsSupport = true;
  generateConstructorsPromptSupport = true;
  generateDelegateMethodsPromptSupport = true;
  moveRefactoringSupport = true;
}
```


Some of these can result in additional items in the `textDocument/codeAction` response. The server is then *not* able to handle the CodeAction.command if it were executed via `workspace/executeCommand` but instead relies on the client knowing how to interpret the command itself.

I'm not sure if there are any plugins yet which use this - for nvim-jdtls there is a PR up https://github.com/mfussenegger/nvim-jdtls/pull/98 that I'll merge soon

So it is not that a plugin can enrich the options provided by the server with additional entries, but instead a plugin can handle specific actions it knows about on the client side.

**Comment by weilbith** [comment]
Ah okay, I see. Thanks for explaining this. So I'll see when there are server + plugin setups which make use of it and then implement it into my plugin.

**Pull request merged**
